### ISSUE TITLE
Redesign

### ISSUE THREAD

**Issue by MystK** [MystK/delta-bot-three#226]
I think DB3 has evolved over time from a simple "check comments" bot to much more.

It does many things now:
Check comments for deltas
Second check comments for deltas
Rescan Edited Comments
Rescan shadow edited comments
User History Page
Yearly Deltaboard
Weekly Deltaboard
Monthly Deltaboard
First time delta PMs
Delta logs
Sticky deltas awarded comment
Mod commands

From looking at the code you can see the evolution, and not everything has been updated to the "latest architecture"

This ticket will be a redesign of DB3 for all of the above features to have the "latest architecture." Some things I can think of are:
Modules
100% snoowrap
Use levelup DB in all needed places

**Comment by hallidev** [comment]
I needed some practice setting up an enterprise quality solution in .NET Core and needed something to build, so purely as an exercise I started building "DeltaBot 4". Once I got to about 25% (maybe less) of DB3's current functionality, I felt I'd done what I set out to and set it aside.

I've made the repo public, feel free to take a look. If nothing else it might give you some perspective on a potential redesign:

https://github.com/hallidev/delta-bot-four

**Comment by MystK** [comment]
@hallidev this is pretty great. Would you be willing to handle deltaboards? One of the biggest issues we have right now is correctly handling deltaboards, the monthly, yearly, and all time specifically. At the moment they are just a chron job so they can easily be disabled on DB3 and moved to DB4.

**Comment by hallidev** [comment]
@MystK So deltaboards are kind of where I left off since I hadn't built in a persistence mechanism like how you're using levelup DB. There are already things I built into this codebase that would be better served with a small DB.

I'd be happy to contribute with a deltaboard redesign along with bringing in a proper local db of some sort. I actually enjoyed building this out, so if there's interest in moving it forward, I'm down with that. I'd need some help though. I'm currently a full time developer with the bright idea of building a game on the side, and I need a few hours here and there where I'm not coding :)

Doing a feature complete rewrite is probably beyond the scope of what I could do on my own, at least in any reasonable amount of time. As for deltaboards and a persistence layer, I could start movement on that as long as there was buy-in from the CMV devs on moving to this solution (.NET Core, C#).

**Comment by Snorrrlax** [comment]
From a mod POV, I will defer to @MystK's judgement on this. Perhaps @hacksoncode could offer some thoughts, but I'm afraid I'm not sure what any of this means :P 

All I know is, if this moves things closer to a robust, complete DeltaBot - I'm happy! 

Also, CMVModBot is written in C#, so my gut says it might be good to align in that regard? https://github.com/perezdev/CMVModBot (cc: @perezdev)

**Comment by MystK** [comment]
In my opinion I don't think it matters what language it's built in, as long as it works. I actually do 50/50 .NET core/JavaScript development at work, so I can assist you, but I'd prefer that you take the lead.

For some insight into how our current deltaboards work, we actually don't use a DB. For the weekly and monthly deltaboards, we look through DeltaBot's comments and create the board from scratch every 30 minutes. The actual call to grab the comments are very quick, a few seconds at most.

For the yearly deltaboards, we use reddit's Amazon elastic timestamp search capabilities to look at all the threads of the year and compile the deltaboards. This has a downside of not being able to grab the deleted threads.

The monthly deltaboards solution will need to move to the yearly deltaboards style as parsing DeltaBot's comments doesn't retrieve the whole month's deltas anymore.

The whole reasoning behind not using a database because keeping that data synced between the DB and reddit and very challenging, but that is my opinion and I am open to trying out new things.

**Comment by Snorrrlax** [comment]
My only concern is the warnings you hear about rewriting code. This will be our fourth attempt (technically fifth) - will we finish it?

Also tagging @wyantb due to significant contributions

**Comment by MystK** [comment]
We probably shouldn't thinking about it as a rewrite though. DeltaBot, in my opinion, is actually a combination of "modules" like:
Handling deltaboards
Handling awarding comments
Handling updating flairs
Handling wiki pages
Handling delta logs
Handling mod commands
Handling edited comments

These modules share sort of similar functionality, but in actually they can be completely isolated environments.

So @hallidev's code is actually just a "module" to handle deltaboards, but it can live with/coincide with the modules that live in this repo.

We could rename this repo to deltabot modules, and @hallidev's to deltabot deltaboards module.

**Comment by Snorrrlax** [comment]
Well, I'm sold! Thoughts, @MystK?

**Comment by MystK** [comment]
Agreed, everything sounds great. @hallidev, perhaps what we could do is a piece by piece migration to DB4 and we could start with deltaboards? 

DB3 separates handling mod commands and put that also on a chron job. Perhaps you can handle that next?

Regarding the rest of the modules, DB3 does all of them together, so it can't be easily split and would have to be upgraded together. 

What do you think @hallidev? Maybe it would be easier to implement "Handling delta logs" first instead and replacing the main DB3 modules first instead?

For DB4, are you adding hidden JSON data to the comments like DB3?

**Comment by hallidev** [comment]
All makes sense to me. I'm handling the delta boards as a command line arg sent to a console app meant to be called from a cron job. They will definitely be a stand alone feature. 

Regarding the hidden JSON data, everything I've built so far is a direct port of the existing code, so it reads and writes all the same stuff in the same format.

**Comment by wyantb** [comment]
Cool! Actual pub/sub mechanisms would go a long way to help individual bits and pieces of logic to stay sane, and type safety is always handy as well. It's been a few years since I've written C# but everything I looked at real quick is plenty legible in my book. Slap some (more?) tests on there for the service implementations and that could serve as a really excellent platform as it grows.

Am I reading right that you're using `RedditSharp` as a reddit API wrapper in there? Do you think the project would run under mono as well as on a windows system? (Not all of us have a windows install - I've been off it for years haha)

**Comment by hallidev** [comment]
Everything is .NET Core 2, so it shouldn't be an issue to open everything with your preferred environment. It should be able to run anywhere. I'm using Visual Studio so there's a .sln file, etc., but it's definitely not required. 

Regarding RedditSharp - you bring up a good point. This API wrapper is pretty well built and has nice features, but the maintenance is spotty and I had some issues when building my codebase out. I'm actually referencing a local copy because I had to clone the repo, make fixes and build it. The mistakes in there at the time (this was back in October) were pretty blatant, so it worried me a bit. It's definitely still a risk, but the repo does appear to be active still. Without abstracting RedditSharp away 100%, I tried to keep it pretty loose. Ripping out RedditSharp and replacing it with something else wouldn't be a daunting task, but yeah it's a concern of mine.

For now I'm only going to focus on a solid deltaboard implementation. If it tests out well and actually gets used on the sub reliably we can take it from there.

This weekend didn't turn out so great for me time wise. I made the console changes necessary to run in "deltaboard" mode and got my environment setup with new accounts, etc. I'm ready to go on it, but didn't get into any serious implementation. I'll put some updates on Slack as I can some time here and there to build this out.

**Comment by Snorrrlax** [comment]
A few questions:

Would it be worth renaming this? Something about "DeltaBot v4" suggests we need new versions. Something that sounds more final might help it _be_ final.

Speaking of which, could we work to milestones/a timeline on this? 

Will the deltaboard module you're working on just now update the wiki page (daily, weekly, monthly, yearly, all-time) as well as the sidebar?

**Comment by hallidev** [comment]
The initial work will be the daily / weekly / monthly / yearly / all-time. I'll need to take a look at the sidebar, but if it's supposed to be updated as part of the delta board process I'll make sure it's included. 

Regarding the naming, it's currently named "DeltaBotFour". We should probably do this deltaboard stuff first before going all in with it though. I'm open to whatever people want though.

**Comment by Snorrrlax** [comment]
Thanks! Yeah, the sidebar should be included in this module, and it's currently having issues: https://github.com/MystK/delta-bot-three/issues/232

**Comment by Snorrrlax** [comment]
Any progress on this?

**Comment by hallidev** [comment]
I've been focused (which was one of my concerns above) with a project that was already active when I mentioned this, but I did get some work into it. I still feel like I'll need a few weeks before I wrap up with my active project and can dedicate time to this.

I did stub out the implementation however. It's now just a question of filling in the blanks when it comes to the Deltaboard work. For someone who knows the current implementation of Deltaboard well, I don't think it will be a big job. For me, 90% of the work is reverse engineering the current solution.

If anyone is willing to contribute, I can get them started and show just where the "work goes here" stuff is. I spent time making sure it would be easy for someone to get going.

If nobody can contribute I understand, but I will need a bit of time. I'm still happy to do it though. Or happy to help anyone out that can contribute sooner.
